src/Plugin/resource/Resource.php

```diff
@@ -567,4 +568,48 @@ protected function initAuthenticationManager() {
     }
   }
 
+  /**
+   * Adds the Allowed-Origin headers.
+   *
+   * @param string $path
+   *   The requested path.
+   */
+  protected function preflight($path) {
+    $plugin_definition = $this->getPluginDefinition();
+    $allowed_origin = empty($plugin_definition['allowOrigin']) ? variable_get('restful_allowed_origin', NULL) : $plugin_definition['allow_origin'];
+    // Always add the allow origin if configured.
+    if ($allowed_origin) {
+      restful()
+        ->getResponse()
+        ->getHeaders()
+        ->add(HttpHeader::create('Access-Control-Allow-Origin', check_plain($allowed_origin)));
+    }
+
+    // Make sure the Access-Control-Allow-Methods is populated.
+    $allowed_methods = array();
+    foreach ($this->getControllers() as $pattern => $controllers) {
+      // Find the controllers for the provided path.
+      if ($pattern == $path || ($pattern && preg_match('/' . $pattern . '/', $path))) {
```

Comment:
but wouldn't you expect the pattern match/access to happen pre-preflight? i.e. we shouldn't get to this point if there's no access.